### PR TITLE
[web] roll CanvasKit 0.36.1 (attempt 3)

### PR DESCRIPTION
**Pull request by yjbanov**
This roll was reverted in https://github.com/flutter/engine/pull/35837 because of golden failures in the engine=>framework roll, which turned out to be improvements. The new version of CK renders more consistently with non-web versions of Flutter ([examples](https://flutter-gold.skia.org/search?corpus=flutter&include_ignored=false&left_filter=name%3Dcupertino.cupertino_magnifier.position.default&max_rgba=0&min_rgba=0&negative=true&not_at_head=false&positive=true&reference_image_required=false&right_filter=&sort=descending&untriaged=true)).

Roll CanvasKit to 0.36.1. This brings a fix for certain old Android GPUs that trigger a bug when used through WebGL.

Fixes https://github.com/flutter/flutter/issues/74990
Fixes https://github.com/flutter/flutter/issues/110684

### CODE REVIEW

File: lib/web_ui/test/canvaskit/canvaskit_api_test.dart [hunk 1]
```diff
@@ -1513,51 +1522,93 @@ void _paragraphTests() {
...
+        'paragraph_kitchen_sink.png',
+        region: const ui.Rect.fromLTRB(0, 0, 400, 160),
+        maxDiffRatePercent: 0.0,
+        write: true,
+      );
+    }
+
+    void expectAlmost(double actual, double expected) {
+      expect(actual, within<double>(distance: actual / 100, from: expected));
+    }
```
Comment (ditman) [inline lib/web_ui/test/canvaskit/canvaskit_api_test.dart:1580]: Do you need your own implementation to validate this?

There's already [`moreOrLessEquals`](https://api.flutter.dev/flutter/flutter_test/moreOrLessEquals.html) and [`closeTo`](https://api.flutter.dev/flutter/package-matcher_matcher/closeTo.html) that might help here (and probably more!).

```dart
expect(actual, moreOrLessEquals(expected, epsilon: actual/100));
```

(Also not sure about using `actual` as part of the epsilon value... shouldn't it be fixed to guarantee accuracy to X decimal points? Like `0.1` or maybe `1`?)

File: lib/web_ui/test/canvaskit/canvaskit_api_test.dart [hunk 2]
```diff
@@ -1513,51 +1522,93 @@ void _paragraphTests() {
...
-    expect(paragraph.getLongestLine(), 50);
-    expect(paragraph.getMaxIntrinsicWidth(), 50);
-    expect(paragraph.getMinIntrinsicWidth(), 50);
-    expect(paragraph.getMaxWidth(), 55);
+    expectAlmost(paragraph.getHeight(), 108);
+    expectAlmost(paragraph.getIdeographicBaseline(), 108);
+    expectAlmost(paragraph.getLongestLine(), 263);
+    expectAlmost(paragraph.getMaxIntrinsicWidth(), 263);
+    expectAlmost(paragraph.getMinIntrinsicWidth(), 135);
+    expectAlmost(paragraph.getMaxWidth(), 500);
```
Comment (ditman) [inline lib/web_ui/test/canvaskit/canvaskit_api_test.dart:1589]: In this case, the width can be `+-5` of `500` (I assume pixels?). Isn't that delta starting to be "too large" that could start masking actual measurement/rounding errors?

File: lib/web_ui/test/canvaskit/canvaskit_api_test.dart [hunk 3]
```diff
@@ -1447,31 +1450,36 @@ void _textStyleTests() {
...
+      ..backgroundColor = Float32List.fromList(<double>[0.2, 0, 0, 0.5])
+      ..color = Float32List.fromList(<double>[0, 1, 0, 1])
+      ..foregroundColor = Float32List.fromList(<double>[1, 0, 1, 1])
       ..decoration = 0x2
       ..decorationThickness = 2.0
       ..decorationColor = Float32List.fromList(<double>[13, 14, 15, 16])
       ..decorationStyle = canvasKit.DecorationStyle.Dotted
       ..textBaseline = canvasKit.TextBaseline.Ideographic
-      ..fontSize = 24
+      ..fontSize = 48
```
Comment (ditman) [inline lib/web_ui/test/canvaskit/canvaskit_api_test.dart:1479]: Why these changes in fontSize, heightMultiplier and others? (I understand the colors were maybe invalid before, but the rest?)

**Review by ditman — APPROVED** [review]
Small nitpicks/questions.

File: lib/web_ui/test/canvaskit/canvaskit_api_test.dart (hunk 1 above)
Comment (yjbanov) [reply lib/web_ui/test/canvaskit/canvaskit_api_test.dart:1580]: `moreOrLessEquals` is from `package:flutter_test`, which we don't have access to. `closeTo` is interesting. Right now `within` is standard in the code base, so I think for now I'll prefer to continue relying on it.

File: lib/web_ui/test/canvaskit/canvaskit_api_test.dart (hunk 2 above)
Comment (yjbanov) [reply lib/web_ui/test/canvaskit/canvaskit_api_test.dart:1589]: This whole test is just a smoke test. More precise text measurement tests are in dedicated files. So I'm not worried about the precision here. TBH I'd probably be happy if we just checked that the returned value is not null. That we are also checking for a specific value is a bonus.

File: lib/web_ui/test/canvaskit/canvaskit_api_test.dart (hunk 3 above)
Comment (yjbanov) [reply lib/web_ui/test/canvaskit/canvaskit_api_test.dart:1479]: I made this test a golden test. Previously the text we rendered was complete nonsense. Now it is displaying something reasonable: https://flutter-engine-gold.skia.org/diff?test=paragraph_kitchen_sink&left=c8f1c90d490826fdbcaf4e2a4762fc76&right=eb3fc116f84bdf3e401fe0345ccde4f3

File: lib/web_ui/dev/canvaskit_lock.yaml [hunk 4]
```diff
@@ -1,4 +1,4 @@
 # Specifies the version of CanvasKit to use for Flutter Web apps.
 #
 # See `lib/web_ui/README.md` for how to update this file.
-canvaskit_version: "0.35.0"
+canvaskit_version: "0.36.1"
```
Comment (loic-sharma) [inline lib/web_ui/dev/canvaskit_lock.yaml:4]: What are your thoughts on rolling to 0.37.0? https://www.npmjs.com/package/canvaskit-wasm/v/0.37.0

Or is the idea to do the minimal update to minimize potential breaks?

Comment (yjbanov) [reply lib/web_ui/dev/canvaskit_lock.yaml:4]: Yeah, I started a draft - https://github.com/flutter/engine/pull/36011 - but I want to isolate the golden flakiness we're seeing into separate rolls. I'm not sure if 0.37.0 could contain other surprises, such as unfixing the issue that 0.36.1 is fixing or some such.

Comment (loic-sharma) [reply lib/web_ui/dev/canvaskit_lock.yaml:4]: Makes sense! Marking as resolved
